Committee expiration, validation and modification

teodanciu opened the pull request:
# Description

<!-- Add your description here, if it fixes a particular issue please provide a
[link](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword=)
to the issue. -->

Closes https://github.com/input-output-hk/cardano-ledger/issues/3443

* In Gov rule, check the `NewCommittee` action failing if : 
   * the members to be removed are not part of the current committee (for this, the current committee from EnactState had to be added to GovEnv) 
   * the expiration epoch of the new members are not exceeding the ppCommitteeTermLimitL, given the current epoch
* Rename `NewCommittee` to `UpdateCommittee` and change its semantics to  "update committee", rather than "replace committee".  This means: in Enact rule: adding and removing members correctly, and in Gov rule: checking correctly the minimum committee size, considering the current committee.
* Update CommitteeState when members get removed

# Checklist

- [ ] Commit sequence broadly makes sense and commits have useful messages
- [ ] New tests are added if needed and existing tests are updated
- [ ] When applicable, versions are updated in `.cabal` and `CHANGELOG.md` files according to the
      [versioning process](https://github.com/input-output-hk/cardano-ledger/blob/master/RELEASING.md#versioning-process).
- [ ] The version bounds in `.cabal` files for all affected packages are updated. **If you change the bounds in a cabal file, that package itself must have a version increase.** (See [RELEASING.md](https://github.com/input-output-hk/cardano-ledger/blob/master/RELEASING.md#versioning-process))
- [ ] All visible changes are prepended to the latest section of a `CHANGELOG.md` for the affected packages. **New section is never added with the code changes.** (See [RELEASING.md](https://github.com/input-output-hk/cardano-ledger/blob/master/RELEASING.md#changelogmd))
- [x] Code is formatted with [`fourmolu`](https://github.com/fourmolu/fourmolu) (use `scripts/fourmolize.sh`)
- [x] Cabal files are formatted (use `scripts/cabal-format.sh`)
- [x] [`hie.yaml`](https://github.com/input-output-hk/cardano-ledger/blob/master/hie.yaml) has been updated (use `scripts/gen-hie.sh`)
- [ ] Self-reviewed the diff

teodanciu committed:
Check that committe members to add are not already expired

otherwise fail with a new predicate failure

teodanciu committed:
Check that committee members to add and the ones to remove are disjoint

otherwise fail with a new PredicateFailure

teodanciu committed:
Rename `NewCommittee` and change the meaning to 'update'

instead of 'replace'

teodanciu commented on eras/conway/impl/src/Cardano/Ledger/Conway/Governance/Procedures.hs:
```diff
@@ -565,8 +567,8 @@ instance EraPParams era => EncCBOR (GovAction era) where
         Sum NoConfidence 3 !> E (encodeNullStrictMaybe encCBOR) gid
-      NewCommittee gid old new ->
-        Sum NewCommittee 4 !> E (encodeNullStrictMaybe encCBOR) gid !> To old !> To new
+      NewCommittee gid old new q ->
```
Could we change the name as well? to `UpdateCommitte`, since we're changing the cddl anyway?  Would that cause too big discrepancies with what is already integrated?

lehins replied:
I think it is a great idea. Please, go ahead, change it in the cddl spec as well as Haskell code. This will reflect better the actual semantics that are described in the CIP-1694

teodanciu replied:
I changed it without deprecation, I suppose they will have to reset sanchonet if this action was already submitted, but I assume that's find at this point

lehins replied:
Yes, we are allowed to break SanchoNet :grin:

teodanciu replied:
:fire:

teodanciu committed:
Update CommitteeState to remove members that were voted out

The udpated state is set both in vState and the govSnapshots of the
returned Epoch state

lehins commented on eras/conway/impl/src/Cardano/Ledger/Conway/Rules/Epoch.hs:
```diff
@@ -336,3 +352,9 @@ instance
+
+updateCommitteeState :: StrictMaybe (Committee era) -> CommitteeState era -> CommitteeState era
+updateCommitteeState committee (CommitteeState creds) =
+  CommitteeState $ Map.restrictKeys creds (Map.keysSet members)
```
No need to convert it to a Set
```suggestion
  CommitteeState $ Map.difference creds members
```

teodanciu replied:
I need to remember that the value type don't have to coincide for `Map.difference` ! Thanks

lehins commented on eras/conway/impl/src/Cardano/Ledger/Conway/Rules/Epoch.hs:
```diff
@@ -336,3 +352,9 @@ instance
+updateCommitteeState committee (CommitteeState creds) =
+  CommitteeState $ Map.restrictKeys creds (Map.keysSet members)
+  where
+    members = foldMap' committeeMembers committee
```
Very cute way of converting Nothing to an empty Map :slightly_smiling_face: :heart:

lehins commented on eras/conway/impl/src/Cardano/Ledger/Conway/Rules/Gov.hs:
```diff
@@ -232,17 +258,32 @@ govTransition = do
-                    ?! NewCommitteeSizeTooSmall newCommitteeSize minCommitteeSize
+          UpdateCommittee _mPrevGovActionId membersToRemove membersToAdd qrm -> do
+            _ <- checkCommitteeSize
+            _ <- checkUnexpectedMembers
```
These two actions return unit, so ignoring output should not be necessary?
```suggestion
            checkCommitteeSize
            checkUnexpectedMembers
```

teodanciu replied:
Oops, scala!

lehins commented on eras/conway/impl/src/Cardano/Ledger/Conway/Rules/Gov.hs:
```diff
@@ -232,17 +258,32 @@ govTransition = do
+              checkExpirationEpoch =
+                let maxTermLimit = pp ^. ppCommitteeTermLimitL
+                    invalidMembers =
+                      Map.filter (> EpochNo (fromIntegral maxTermLimit) + currentEpoch) $ membersToAdd
```
This check is too restrictive, since it will take a few epochs to vote on such proposal. In the matter of fact, I would prefer for the `UpdateCommittee` action to contain number of epochs that a committee member will be active for, rather than the absolute Epoch number of when a committee member expires:

@WhatisRT I couldn't find in the spec how we enforce [`ccTermLimit`](https://github.com/input-output-hk/formal-ledger-specifications/blob/7e2e805a9b4db18e73fff75aa6865d65213f004d/src/Ledger/PParams.lagda#L74), but that's how I would implement it. Does this align with your plans?

Also, redundant `$` and `EpochNo` has a `Num` instance, so this can be simplified a bit:
```suggestion
                      Map.filter (> fromIntegral maxTermLimit) membersToAdd
```

WhatisRT replied:
I agree that it's a little awkward that it's an absolute epoch instead of a number of epochs, but it is that way for a good reason: you want full predictability here. For example, if there's going to be a schedule where one third of the CC is replaced each year it's much better to align the schedule exactly, instead of having a few weeks where some time out but others don't.

Two notes about `ccTermLimit`:
- First, in my PR to the CIP I've renamed it to `ccMaxTermLength`, because that's what it actually is.
- My plan is to enforce it in `ENACT`. The reason is that this check can become successful over time, and we only know when it'll be successful in `ENACT`.

lehins replied:
> if there's going to be a schedule where one third of the CC is replaced each year it's much better to align the schedule exactly

Wouldn't it be as easy as submitting an `UpdateCommittee` GovAction that only updates the lifetime of 1/3 of the committee to the same duration?

lehins commented on eras/conway/impl/src/Cardano/Ledger/Conway/Rules/Gov.hs:
```diff
@@ -262,3 +303,19 @@ govTransition = do
+  Map.Map (Credential 'ColdCommitteeRole (EraCrypto era)) EpochNo ->
+  UnitInterval ->
+  StrictMaybe (Committee era) ->
+  StrictMaybe (Committee era)
```
This function always returns new committee, so if SNOthing case is not possible, this function should not have the `StrictMaybe` return type.
```suggestion
  Committee era
```

teodanciu replied:
💯

lehins commented on eras/conway/impl/src/Cardano/Ledger/Conway/Rules/Gov.hs:
```diff
@@ -262,3 +303,19 @@ govTransition = do
+        ( Committee
+            (currentMembers `Map.withoutKeys` membersToRemove `Map.union` membersToAdd)
+            newQuorum
+        )
```
Firstly, `union` is left biased, therefore we want to use the new EpochNo from the `membersToAdd`, rather than the values from the current state of `currentMembers`
Secondly, using a let binding improves formatting a bit :slightly_smiling_face: 
```suggestion
      let newCommitteeMembers = 
            Map.union membersToAdd (currentMembers `Map.withoutKeys` membersToRemove)
        in Committee newCommitteeMembers newQuorum
```

teodanciu replied:
Eeek!  Thank you

lehins commented on eras/conway/impl/src/Cardano/Ledger/Conway/Rules/Gov.hs:
```diff
@@ -232,17 +258,32 @@ govTransition = do
+                let maxTermLimit = pp ^. ppCommitteeTermLimitL
+                    invalidMembers =
+                      Map.filter (> EpochNo (fromIntegral maxTermLimit) + currentEpoch) $ membersToAdd
+                 in Map.null invalidMembers ?! InvalidCommitteeExpirationEpoch invalidMembers
```
It's good to expected value of the term limit to the predicate failure. Also renaming it could make sense to reflect the semantics a bit better
```suggestion
                 in Map.null invalidMembers ?! InvalidCommitteeTermLimit maxTermLimit invalidMembers
```

lehins commented on eras/conway/impl/src/Cardano/Ledger/Conway/Rules/Gov.hs:
```diff
@@ -232,17 +288,32 @@ govTransition = do
+                    committeeSize = fromIntegral . Map.size $ foldMap' (^. committeeMembersL) newCommittee
+                 in committeeSize
+                      >= minCommitteeSize
+                        ?! NewCommitteeSizeTooSmall committeeSize minCommitteeSize
```
I do like this check, however we can't apply it this way, because there is a chance, that there is a previous `UpdateCommitte` governance action, that this one relies on, which haven't been enacted yet.

We would only be able to implement this check if we also implement #3655 and resolve all of the proposals that are currently in-flight.

lehins commented on eras/conway/impl/src/Cardano/Ledger/Conway/Rules/Enact.hs:
```diff
@@ -86,10 +87,10 @@ enactmentTransition = do
       pure $
         st
-          & ensCommitteeL .~ SJust committee -- TODO: check old members
+          & ensCommitteeL %~ updatedCommittee membersToRemove membersToAdd newQuorum
```
We need to do all the checks (expiration and members to remove) during ratification. In other words, predicate checks during proposal in GOV rule do not guarantee that upon enactment they are still valid. Here are reasons why they might become invalid:

* `ppCommitteeTermLimitL` could have been updated between proposal and unactment
* another `UpdateCommittee` got enacted before this one, which affected the current committee state. (I've mentioned this case in another comment of mine)

I would suggest creating `verifyUpdateCommittee` and a custom failure type that indicates what is wrong, which could then potentially be used in the GOV rule as a predicate failure, as well as in RATIFY rule to prevent invalid `UpdateCommittee`. This whole logic could potentially live in a single `updateCommittee` function, if that would improve performance, but that is definitely not a requirement for this GOvAction, especially not for the initial version of it.

lehins commented on eras/conway/impl/src/Cardano/Ledger/Conway/Rules/Epoch.hs:
```diff
@@ -302,6 +316,8 @@ epochTransition = do
       & esGovernanceL . cgGovSnapshotsL .~ newGov
       -- Move donations to treasury
       & esAccountStateL . asTreasuryL <>~ donations
+      -- remove hot keys of commitee members that were removed
```
```suggestion
      -- remove hot keys of committee members that were removed
```

lehins commented on eras/conway/impl/src/Cardano/Ledger/Conway/Rules/Epoch.hs:
```diff
@@ -289,9 +304,8 @@ epochTransition = do
-            es'
-              ^. esLStateL . lsCertStateL . certVStateL . vsCommitteeStateL
+        , -- committee state snapshot will not contain members that were removed in Enact
+          prevCommitteeState = updatedCommitteeState
```
I believe previous implementation was actually correct. We need to retain the committee state from the previous epoch, in order for committee votes to be valid
```suggestion
        , prevCommitteeState =
            es'
              ^. esLStateL . lsCertStateL . certVStateL . vsCommitteeStateL
```

If I am wrong we'll fix it later, since it should match the DRep implementation above (which is currently unused).


In any case we'll reevaluate the whole time travel thing later, especially after #3719 is merged. For now I think it would be best to revert the implementation, even if I am wrong in my understanding of this flow.

teodanciu replied:
It might actually not even matter so much, since it's only removing information that shouldn't be used anyway  (since the vote of the removed members shouldn't even get to be looked up in this map), but  I agree, better to do less if unsure and add to it later :+1:

lehins reviewed (changes requested):
This is a very nice piece of work. Some fixes are needed, but overall looks great. 

Also needs a rebase on master.

WhatisRT commented:
`NewCommittee` shouldn't fail in `GOV` if the removed members aren't in the CC. The reason for this is simple: you only know the current `EnactState`, but not necessarily the one that this GA is actually supposed to apply to. You could argue that this logic should then go into `ENACT`, and I originally wanted to put it there, but I've removed it again. The reasoning is here: https://github.com/input-output-hk/formal-ledger-specifications/pull/206#discussion_r1322786893

Long story short, I think we should only do this check if we keep track of in-flight GAs & validate new ones against the current state and the ones in the queue.

lehins commented:
> NewCommittee shouldn't fail in GOV if the removed members aren't in the CC. The reason for this is simple: you only know the current EnactState, but not necessarily the one that this GA is actually supposed to apply to.

@WhatisRT Yes, I brought up the same point here: https://github.com/input-output-hk/cardano-ledger/pull/3721#discussion_r1323240959

One more reason in favor of #3655

teodanciu committed:
Move committee size validation to `Ratify`

so that we check against the relevant version of the current committe

teodanciu committed:
Delay ratification if the expiry of any committee member is too big

relative to the `committeeTermLimitL` pparam.

teodanciu commented:
@lehins I think the latest covers the items that came up after the meeting yesterday, if you would like to take another look. 

* Prevent expiry epochNo less or equal to current epochNo in GOV
* Accept expiry from the past on Enactment
* Delay expiry that is too long
* Prevent intersection of removed and added/updated in GOV

lehins reviewed (approved):
Beautiful! :+1:
